### PR TITLE
Prevent `v_oc` to be negative in `_lambertw`

### LINKED ISSUE #1673: singlediode error with very low effective_irradiance
**Describe the bug**

Since pvlib 0.9.4 release (https://github.com/pvlib/pvlib-python/pull/1606) I get an error while running the single-diode model with some very low effective irradiance values.

**To Reproduce**

```python
from pvlib import pvsystem

effective_irradiance=1.341083e-17
temp_cell=13.7 

cec_modules = pvsystem.retrieve_sam('CECMod')
cec_module = cec_modules['Trina_Solar_TSM_300DEG5C_07_II_']

mount = pvsystem.FixedMount()
array = pvsystem.Array(mount=mount,
                       module_parameters=cec_module)

system = pvsystem.PVSystem(arrays=[array])

params = system.calcparams_cec(effective_irradiance, 
                               temp_cell)

system.singlediode(*params)
```

```in _golden_sect_DataFrame(params, lower, upper, func, atol)
    303 """
    304 Vectorized golden section search for finding maximum of a function of a
    305 single variable.
   (...)
    342 pvlib.singlediode._pwr_optfcn
    343 """
    344 if np.any(upper - lower < 0.):
--> 345     raise ValueError('upper >= lower is required')
    347 phim1 = (np.sqrt(5) - 1) / 2
    349 df = params

ValueError: upper >= lower is required
```

**Expected behavior**
This complicates the bifacial modeling procedure as `run_model_from_effective_irradiance` can be called with very low irradiance values estimated by pvfactors (at sunrise or sunset for instance). 

**Versions:**
 - ``pvlib.__version__``:  0.9.4
 - ``pandas.__version__``: 1.5.3
 - python: 3.10

**Additional context**

v_oc is negative in this case which causes the error. 

```python
from pvlib.singlediode import _lambertw_v_from_i
photocurrent = params[0]
saturation_current = params[1]
resistance_series = params[2]
resistance_shunt = params[3]
nNsVth = params[4]
v_oc = _lambertw_v_from_i(resistance_shunt, resistance_series, nNsVth, 0.,
                              saturation_current, photocurrent)
```

### LINKED ISSUE #1780: _golden_sect_DataFrame changes in 0.9.4
**Describe the bug**

`0.9.4` introduced the following changes in the `_golden_sect_DataFrame`: We are checking `upper` and `lower` parameters and raise an error if `lower > upper`.

https://github.com/pvlib/pvlib-python/blob/81598e4fa8a9bd8fadaa7544136579c44885b3d1/pvlib/tools.py#L344-L345

`_golden_sect_DataFrame` is used by `_lambertw`:

https://github.com/pvlib/pvlib-python/blob/81598e4fa8a9bd8fadaa7544136579c44885b3d1/pvlib/singlediode.py#L644-L649

I often have slightly negative `v_oc` values (really close to 0) when running simulations (second number in the array below):
```
array([ 9.46949758e-16, -8.43546518e-15,  2.61042547e-15,  3.82769773e-15,
        1.01292315e-15,  4.81308106e+01,  5.12484772e+01,  5.22675087e+01,
        5.20708941e+01,  5.16481028e+01,  5.12364071e+01,  5.09209060e+01,
        5.09076598e+01,  5.10187680e+01,  5.11328118e+01,  5.13997628e+01,
        5.15121386e+01,  5.05621451e+01,  4.80488068e+01,  7.18224446e-15,
        1.21386700e-14,  6.40136698e-16,  4.36081007e-16,  6.51236255e-15])
```

If we have one negative number in a large timeseries, the simulation will crash which seems too strict.

**Expected behavior**

That would be great to either:
* Have this data check be less strict and allow for slightly negative numbers, which are not going to affect the quality of the results.
* On `_lambertw`: Do not allow negative `v_oc` and set negative values to `np.nan`, so that the error is not triggered. It will be up to the upstream code (user) to manage those `np.nan`.

**Versions:**
 - ``pvlib.__version__``: >= 0.9.4
 - ``pandas.__version__``: 1.5.3
 - python: 3.10.11

### PR DESCRIPTION
**Pull request by cedricleroy**
- [x] Closes #1780, closes #1673
 - [x] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [x] Tests added
 - [x] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [x] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [x] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [x] Pull request is nearly complete and ready for detailed review.
 - [x] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

### CODE REVIEW

File: pvlib/singlediode.py [hunk 1]
```diff
@@ -794,6 +794,13 @@ def _lambertw(photocurrent, saturation_current, resistance_series,
     # Compute open circuit voltage
     v_oc = _lambertw_v_from_i(0., **params)
 
+    # Set elements <0 or close to 0 in v_oc to 0
+    if isinstance(v_oc, np.ndarray):
+        v_oc[(v_oc < 0) | ((v_oc > 0) & (v_oc < 1e-12))] = 0.
+    elif isinstance(v_oc, (float, int)):
+        if v_oc < 0 or 0 < v_oc < 1e-12:
```
Comment (cwhanse) [inline pvlib/singlediode.py:801]: What I had in mind, is only setting small negative values to zero. Maybe it's caution for its own sake, but I'd prefer to keep larger negatives which must be from some user error or code problem other than roundoff.

```
v_oc[(v_oc <0) & (v_oc > -1e-12)] = 0.
```

File: docs/sphinx/source/whatsnew/v0.10.0.rst [hunk 2]
```diff
@@ -49,6 +49,8 @@ Enhancements
 Bug fixes
 ~~~~~~~~~
 
+* Prevent `v_oc` to be negative in `_lambertw` (:pull:`1782`)
```
Comment (cwhanse) [inline docs/sphinx/source/whatsnew/v0.10.0.rst:52]:
```suggestion
* Prevent small negative values of `v_oc` in :py:func:`pvlib.singlediode._lambertw`
  which result from accumulated roundoff error. (:issue:`1780`, :issue:`1673`, :pull:`1782`)
```

**Review by cwhanse — APPROVED** [review]
@cedricleroy thanks!

@kandersolar your call for v0.10 or later

**Review by kandersolar — APPROVED** [review]
Thanks @cedricleroy!

**Pull request merged**